geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/FileSystemThreadedReader.scala

```diff
@@ -8,101 +8,327 @@
+  *
+  * @param es executor service used by the read tasks
+  * @param phaser phaser for tracking read task completion
+  * @param queue intermediate queue populated by the read tasks
+  */
+class FileSystemThreadedReader private (es: ExecutorService, phaser: Phaser, queue: BlockingQueue[SimpleFeature])
+    extends CloseableIterator[SimpleFeature] {
 
-  def apply(
-      readers: Iterator[(FileSystemPathReader, Iterator[Path])],
-      threads: Int): CloseableIterator[SimpleFeature] = {
-    if (threads < 2) {
-      CloseableIterator(readers).flatMap { case (factory, paths) => CloseableIterator(paths).flatMap(factory.read) }
+  private val localQueue = new java.util.LinkedList[SimpleFeature]()
+
+  private var current: SimpleFeature = _
+
+  override def hasNext: Boolean = {
+    if (current != null) {
+      return true
+    }
+    current = localQueue.pollFirst
+    if (current != null) {
+      return true
+    }
+
+    while (!phaser.isTerminated) {
+      current = queue.poll(100, TimeUnit.MILLISECONDS)
+      if (current != null) {
+        queue.drainTo(localQueue, 10000)
+        return true
+      }
+    }
+    // last check - if phaser.isTerminated, the queue should have whatever values are left
+    current = queue.poll()
+    if (current != null) {
+      queue.drainTo(localQueue, 10000)
+      true
+    } else {
+      false
+    }
+  }
+
+  override def next(): SimpleFeature = {
+    if (hasNext) {
+      val ret = current
+      current = null
+      ret
     } else {
-      new MultiThreadedFileSystemReader(readers, threads)
+      Iterator.empty.next
     }
   }
 
-  class MultiThreadedFileSystemReader(readers: Iterator[(FileSystemPathReader, Iterator[Path])], threads: Int)
-      extends CloseableIterator[SimpleFeature] with StrictLogging {
+  override def close(): Unit = es.shutdownNow()
+}
 
-    private val es = Executors.newFixedThreadPool(threads)
+object FileSystemThreadedReader extends StrictLogging {
 
-    private val queue = new LinkedBlockingQueue[SimpleFeature](2000000)
+  def apply(
+      readers: Iterator[(FileSystemPathReader, Seq[StorageFilePath])],
+      threads: Int): CloseableIterator[SimpleFeature] = {
 
-    private val localQueue = new java.util.LinkedList[SimpleFeature]()
+    if (threads < 2) {
+      readers.flatMap { case (reader, files) =>
+        val mods = scala.collection.mutable.HashSet.empty[String]
+        // ensure files are sorted in reverse chronological order
+        CloseableIterator(files.sorted.iterator).flatMap(f => read(reader, f, mods))
+      }
+    } else {
+      val queue = new LinkedBlockingQueue[SimpleFeature](2000000)
+      val es = Executors.newFixedThreadPool(threads)
 
-    private var current: SimpleFeature = _
+      val phaser = new Phaser(1) {
+        override protected def onAdvance(phase: Int, registeredParties: Int): Boolean = {
+          // when all tasks have been completed, shutdown the executor service
+          es.shutdown()
```

Comment:
As a sanity check, is shutdown the right thing.  Should we awaitTermination or shutdownNow instead?

Reply:
When this gets triggered, all the tasks should be essentially done, or at least in the 'finally' block before exiting, and there shouldn't be any more tasks being submitted, so a regular shutdown should be safe. I don't think we'd want to use shutdownNow, as that might try to interrupt the running tasks (although they should be essentially done, as noted). We could awaitTermination, but this is an asynchronous call so I don't think there's much point in keeping the thread waiting.

Reply:
Awesome.  Thanks for the sanity check!